### ISSUE TITLE
memory leaks 

### ISSUE THREAD

**Issue by alexsanderkhitev** [xmartlabs/Eureka#1000]
Hello! I have a memory leak in Eureka framework. I always use weak links everywhere, while I do not use Form anywhere in the blocks. How can I fix it?

![memory xcworkspace 2017-04-12 16-41-09](https://cloud.githubusercontent.com/assets/12511909/24958074/1c71ef76-1f9f-11e7-81ee-d9cc0598a19c.png)

**Comment by mtnbarreto** [comment]
Hi @alexsanderkhitev I will look into it when i have a chance. 
Could you please share the Eureka form that is causing the memory leak?? Which is the eureka version ?

**Comment by mtnbarreto** [comment]
same as #974

**Comment by alexsanderkhitev** [comment]
@mtnbarreto I can share my code 

I use Eureka (2.0.1)

```
class EventsUpcomingOwnerVC: FormViewController {
    
    fileprivate enum SectionType: String {
        case main, requests, approved, navigation, settings
    }
    
    // MARK: - Data
    
    var card: CardModel!
    
    // MARK: - Delegates
    
    weak var delegate: EventsUpcomingOwnerVCDelegate?
    
    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        setupSettings()
        // ui
        setupUISettings()
        setupNavSettings()
        setupForm()
        // observers
//        firObserve(true)
        debugPrint("FormViewController init")
    }
    
    deinit {
        guard card != nil else { return }
        FIRCardDatabaseManager.shared.observeInterestedRequests(card, isObserve: false, getNewRequest: nil)
        debugPrint("FormViewController deinit")
    }

    private func setupSettings() {
        definesPresentationContext = true
    }
    
    // MARK: - UI
    
    private func setupUISettings() {
        view.backgroundColor = .white
    }
    
    private func setupNavSettings() {
        guard card != nil else { return }
        navigationItem.title = card.title
    }
    
    // MARK: - Form 
    
    private func setupForm() {
        form
            +++ section(.main, headerHeight: nil, footerHeight: nil)
            <<< cardPlacePhotoRow()
            <<< cardInfoRow()
            <<< cardDetailLocationRow()
        setupRequestsSections()
        setupApprovedSection()
        
        let navigationSection = section(.navigation, headerHeight: 30, footerHeight: nil)
        form
            +++ navigationSection
            <<< upcomingChatRow()
        
        let settingsSection = section(.settings, headerHeight: 30, footerHeight: 15)
        form
            +++ settingsSection
            <<< upcomingSettingsRow(.deleteEvent)
        settingsSection.evaluateHidden()
        
    }
    
    private func setupRequestsSections() {
        guard card != nil else { return }
        let requestsSection = section(.requests, headerHeight: 30, footerHeight: nil)
        
        let timeManager = TimeManager()
        card.interestedUsers.sort { (first, second) -> Bool in
            guard let firstDate = timeManager.tranformISOToDate(first.isoDate) else {
                return first.isoDate < second.isoDate
            }
            guard let secondDate = timeManager.tranformISOToDate(second.isoDate) else {
                return first.isoDate < second.isoDate
            }
            
            return firstDate < secondDate
        }
        
        form
            +++ requestsSection
        for requestUserModel in card.interestedUsers {
            if !requestUserModel.isApproved {
                requestsSection
                    <<< upcomingRequestRow(requestUserModel)
            }
        }
    
        requestsSection.evaluateHidden()
    }
 

    private func setupApprovedSection() {
        guard card != nil else { return }
        let approvedSection = section(.approved, headerHeight: 30, footerHeight: nil)
        
        form
            +++ approvedSection
        
        for requestUserModel in card.interestedUsers {
            if requestUserModel.isApproved {
                approvedSection
                    <<< upcomingRequestRow(requestUserModel)
            }
        }
        approvedSection.evaluateHidden()
    }
    
}

// MARK: - Observers

extension EventsUpcomingOwnerVC {
    
    fileprivate func firObserve(_ isObserve: Bool) {
        guard card != nil else { return }
        FIRCardDatabaseManager.shared.observeInterestedRequests(card, isObserve: isObserve) { [weak self] (newCardUserInterestedRequest) in
            guard self != nil else { return }
            guard self!.card.interestedUsers.contains(newCardUserInterestedRequest) != true else { return }
            self!.card.interestedUsers.append(newCardUserInterestedRequest)
            
            guard let requestSection = self!.getSectionBy(.requests) else { return }
            requestSection
                <<< self!.upcomingRequestRow(newCardUserInterestedRequest)
            if requestSection.isHidden {
                requestSection.evaluateHidden()
            }
        }
    }
    
}

// MARK: - Sections

extension EventsUpcomingOwnerVC {
    
    fileprivate func section(_ type: SectionType, headerHeight: CGFloat?, footerHeight: CGFloat?) -> Section {
        let tag = type.rawValue
        let section = Section(tag)
        section.tag = tag
        
        if headerHeight == nil {
            section.header = {
                var header = HeaderFooterView<UIView>(.callback({
                    let view = UIView(frame: CGRect(x: 0, y: 0, width: 0, height: 0))
                    return view
                }))
                header.height = { headerHeight ?? CGFloat.leastNormalMagnitude }
                return header
            }()
        } else {
            switch type {
            case .main:
                break
            case .requests:
                section.header?.title = "Requests"
            case .approved:
                section.header?.title = "Approved"
            case .navigation:
                section.header?.title = "Navigation"
            case .settings:
                section.header?.title = "Settings"
            }
        }
        
        section.footer = {
            var footer = HeaderFooterView<UIView>(.callback({
                let view = UIView(frame: CGRect(x: 0, y: 0, width: 0, height: 0))
                return view
            }))
            footer.height = { footerHeight ?? CGFloat.leastNormalMagnitude }
            return footer
        }()
        
        // hide 
        
        section.hidden = Condition.function([], { (form) -> Bool in
            guard let _section = form.sectionBy(tag: tag) else { return false }
            if _section.count == 0 {
                return true
            } else {
                return false
            }
        })
        
        return section
    }
    
    fileprivate func getSectionBy(_ type: SectionType) -> Section? {
        let tag = type.rawValue
        guard let section = form.sectionBy(tag: tag) else { return nil }
        return section
    }
    
}

// MARK: - Rows

extension EventsUpcomingOwnerVC {
    
    fileprivate func cardPlacePhotoRow() -> CardPlacePhotoRow {
        let tag = "CardPlacePhotoRow"
        let cardPlacePhotoRow = CardPlacePhotoRow(tag: tag)
        guard card != nil else { return cardPlacePhotoRow }
        let value = CardDetailCellModel(cardModel: card)
        cardPlacePhotoRow.value = value
        return cardPlacePhotoRow
    }
    
    fileprivate func cardInfoRow() -> CardInfoRow {
        let tag = "CardInfoRow"
        let row = CardInfoRow(tag: tag)
        guard card != nil else { return row }
        row.value = CardDetailCellModel(cardModel: card)
        
        row.cell.setupCurrentDistance()
        return row
    }
    
    
    fileprivate func cardDetailLocationRow() -> CardDetailLocationRow {
        let tag = "CardDetailLocationRow"
        let row = CardDetailLocationRow(tag: tag)
        guard card != nil else { return row }
        row.value = CardDetailCellModel(cardModel: card)
        return row
    }
    
    fileprivate func upcomingRequestRow(_ cardUserRequestModel: CardUserRequestModel) -> UpcomingRequestRow {
        let tag = cardUserRequestModel.userID
        let row = UpcomingRequestRow(tag: tag)
        guard card != nil else { return row }
        row.value = UpcomingRequestModel(cardModel: card, cardUserRequestModel: cardUserRequestModel)
        row.cell.delegate = self 
        
        return row
    }
    
    fileprivate func upcomingChatRow() -> UpcomingChatRow {
        let tag = "UpcomingChatRow"
        let row = UpcomingChatRow(tag: tag)
        return row
    }
    
    fileprivate func upcomingSettingsRow(_ mode: UpcomingSettingsCell.Mode!) -> UpcomingSettingsRow {
        let tag = "UpcomingSettingsRow"
        let row = UpcomingSettingsRow(tag: tag)
        guard card != nil else { return row }
        row.value = UpcomingSettingsCellModel(mode: mode, card: card)

        row.onCellSelection { [weak self] (cell, row) in
            guard mode != nil else { return }
            switch mode! {
            case .deleteEvent:
                self?.upcomingSettingsRowPresentCardDeleteAlert()
            }
        }
        
        return row
    }
    
    
    // MARK: Getting rows
    
    fileprivate func getCardInfoRow() -> CardInfoRow? {
        let row = form.rowBy(tag: "CardInfoRow") as? CardInfoRow
        return row
    }
    
}

// MARK: - Rows actions 

extension EventsUpcomingOwnerVC {
    
    // MARK: UpcomingSettingsRow
    
    fileprivate func upcomingSettingsRowPresentCardDeleteAlert() {
        guard card != nil else { return }
        let alertController = UIAlertController(title: nil, message: "Delete the card and this event?", preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        
        let okAction = UIAlertAction(title: "OK", style: .default) { [weak self] (action) in
            guard self != nil else { return }
            self!.delegate?.didDeleteCard(self!.card)
            self!.navigationController?.popViewController(animated: true)
            FIRCardDatabaseManager.shared.deleteCard(self!.card, success: {
            }, fail: { (error) in
              
            })
        }
        alertController.addAction(okAction)
        
        present(alertController, animated: true, completion: nil)
    }
    
}

// MARK: - Rows and Cells Delegate 

extension EventsUpcomingOwnerVC: UpcomingRequestCellDelegate {
    
    func switchChangeValue(row: UpcomingRequestRow, isOn: Bool) {
        guard var section = row.section else { return }
        guard let index = section.index(of: row) else { return }
        guard let cardUserRequestModel = row.value?.cardUserRequestModel else { return }
        section.remove(at: index)
        
        if section.count == 0 {
            section.evaluateHidden()
        }
        
        if isOn {
            // approve
            guard let approveSection = getSectionBy(.approved) else { return }
            cardUserRequestModel.isApproved = isOn
            let row = upcomingRequestRow(cardUserRequestModel)
            approveSection
                <<< row
            approveSection.evaluateHidden()
            guard let indexPath = row.indexPath else { return }
            tableView?.scrollToRow(at: indexPath, at: .top, animated: true)
        } else {
            // request
            guard let requestSection = getSectionBy(.requests) else { return }
            cardUserRequestModel.isApproved = isOn
            let row = upcomingRequestRow(cardUserRequestModel)
            requestSection
                <<< row
            requestSection.evaluateHidden()
            guard let indexPath = row.indexPath else { return }
            tableView?.scrollToRow(at: indexPath, at: .top, animated: true)
        }
    }
    
    func addedAdditionalSeat() {
        let cardInfoRow = getCardInfoRow()
        cardInfoRow?.updateCell()
    }
    
}
```

**Comment by m-revetria** [comment]
Hi @alexsanderkhitev, this should be fixed in #1001. Let us know if the issue persist.

Thanks
